blog: `k8s_custom_deploy` intro post

milas opened the pull request:
Short post using a public Kafka Helm chart as an example.

nicks reviewed (commented):
i think this is ok if the goal is to get the announcement out and have a tutorial to point people at

i feel like the high-level narrative is a bit muddled. there's a lot about tilt internals. the example uses Helm. Other places talk about custom tools. I feel like it requires a lot of context to understand. but maybe that's ok.

@Chimney42 what do you think?

Chimney42 commented on blog/_posts/2021-11-15-k8s-custom-deploy.md:
```diff
@@ -0,0 +1,106 @@
+If you've used Tilt, chances are you are well acquainted with the [`k8s_yaml`][api-k8s_yaml] and [`k8s_resource`][api-k8s_resource] Tiltfile functions.
+The [`k8s_yaml`][api-k8s_yaml] function deploys entities to Kubernetes from either a file or string.
+The [`k8s_resource`][api-k8s_resource] function configures port forwards and other behaviors for the deployed entities.
+
```
I feel we're missing the problem statement. Maybe add a sentence like "But sometimes, a kubernetes manifest is not enough. Maybe you want something to manage the entire lifecycle of multiple k8s resources that all belong together, like Helm"

Chimney42 replied:
We should also mention early in this blogpost that even though we use Helm as an example, this is applicable to all kinds of k8s deployment patterns

Chimney42 commented on blog/_posts/2021-11-15-k8s-custom-deploy.md:
```diff
@@ -0,0 +1,106 @@
+Behind the scenes, both of these invoke `helm template` via CLI to get a locally-rendered version of the chart YAML to pass to the [`k8s_yaml`][api-k8s_yaml] built-in Tilt function.
+That is, the actual deployment to Kubernetes is done via Tilt using the Kubernetes API rather than Helm itself.
+
+More generally, this is the pattern we recommend for integrating custom Kubernetes deploy tools with Tilt, and for a lot of local development use cases, this works great!
```
It's a bit confusing that you say "integrating custom K8s deploy tools" while the only pattern I see is to deploy helm. I'd suggest to either change to "for integrating helm deploys", or expand a bit on what pattern you mean (i.e. give more examples how to deploy other than with helm)

Chimney42 commented on blog/_posts/2021-11-15-k8s-custom-deploy.md:
```diff
@@ -0,0 +1,106 @@
+That is, the actual deployment to Kubernetes is done via Tilt using the Kubernetes API rather than Helm itself.
+
+More generally, this is the pattern we recommend for integrating custom Kubernetes deploy tools with Tilt, and for a lot of local development use cases, this works great!
+Tilt can take "shortcuts" to ensure a frictionless and snappy experience because, as a dev tool, its concerns are very different than those of a production deployment tool.
```
What are these shourtcuts? Can you list some examples?

milas replied:
I struck this line entirely - fwiw though, what I was referring to is that Helm creates a full new stack for each release to enable things like easy rollback, so it can be slower than just "in-place" patching of resources. Good for prod, unnecessary for local dev

Chimney42 commented on blog/_posts/2021-11-15-k8s-custom-deploy.md:
```diff
@@ -0,0 +1,106 @@
+This allows tools like Helm to not only delete any entities from Kubernetes that they created but also clean up any extra state (e.g. Helm release metadata).
+
+## What's The Catch?
+We are still exploring the best semantics for passing Tilt-built image references to external tools, so they can interoperate with `docker_build`.
```
I feel like the problem here is not clear enough. I'd recommend adding a sentence about it.
"At the moment, it is still not possible to build an image externally and pass that image reference into the k8s_custom_deploy function"

Chimney42 commented:
> i think this is ok if the goal is to get the announcement out and have a tutorial to point people at
> 
> i feel like the high-level narrative is a bit muddled. there's a lot about tilt internals. the example uses Helm. Other places talk about custom tools. I feel like it requires a lot of context to understand. but maybe that's ok.
> 
> @Chimney42 what do you think?

Good point.
I agree, this blogpost focuses a lot on the technical implementation and a bit too little on the problem it solves, for my taste.
I think, if we can add one or two sentences at the very beginning explaining the problem/use-case it would feel more rounded.

milas commented:
Thanks for the feedback, much appreciated!

I revamped the intro/context, let me know if it reads better now.

I played around with both Knative + OpenFaaS this morning -- I'm going to replace the example with OpenFaaS, which I think is better suited to this as it deploys stuff to K8s via its CLI but doesn't really have YAML template features like Helm/Knative. (Not _totally_ true, but you need to use the operator to get OpenFaaS CRDs, which isn't the default afaict)

milas commented on blog/_posts/2021-11-15-k8s-custom-deploy.md:
```diff
@@ -0,0 +1,106 @@
+---
+
+If you've used Tilt, chances are you are well acquainted with the [`k8s_yaml`][api-k8s_yaml] and [`k8s_resource`][api-k8s_resource] Tiltfile functions.
+The [`k8s_yaml`][api-k8s_yaml] function deploys entities to Kubernetes from either a file or string.
```
Went for "objects" to keep distinct from Tilt resources 🙃

milas commented on blog/_posts/2021-11-15-k8s-custom-deploy.md:
````diff
@@ -0,0 +1,106 @@
+```
+
+Here's what it looks like when we run `tilt up`:
+![Tilt web UI after running "tilt up" against the sample Tiltfile included in this blog post](/assets/images/k8s-custom-deploy/demo.gif)
````
GIF is gone! (Will make sure the merge is a squash so that it gets GC'd and doesn't live in history)

milas commented:
Okay, updated the example to use OpenFaaS which has a `faas-cli deploy` cmd that builds + deploys an image w/o a Dockerfile or any K8s YAML, so fits better into this pattern.

Also added a short section on using `tilt get kapp` to debug

milas commented:
Oh wow @nicks I totally missed tilt-dev/tilt-extensions#297 (knative extension) the other day.

I'm going to add a call-out to that at the end here!

ellenkorbes commented on blog/_posts/2021-11-22-k8s-custom-deploy.md:
```diff
@@ -0,0 +1,123 @@
+Other tools, such as [OpenFaaS][openfaas], provide a streamlined build & deploy experience managed entirely by their CLI, so there's no YAML at all.
+It's also common for teams to have their own custom deployment shell scripts (e.g. for use with CI/CD).
+
+While it's possible to use Tilt's [`local_resource`][api-local_resource] function as a workaround in these instances, that means giving up much of Tilt's deep Kubernetes integration for the resource as a result.
```
> Tilt's deep Kubernetes integration

What does this do?

ellenkorbes commented on blog/_posts/2021-11-22-k8s-custom-deploy.md:
```diff
@@ -0,0 +1,123 @@
+On `tilt down`, our `delete_cmd` will be invoked, which allows `faas-cli` to delete any objects from Kubernetes it created as well as clean up any extra OpenFaaS state.
+
+## What if Something Goes Wrong?
+Integrating an external tool can be tricky!
```
```suggestion
Integrating your workflow with an external tool can be tricky!
```

ellenkorbes commented on blog/_posts/2021-11-22-k8s-custom-deploy.md:
```diff
@@ -0,0 +1,123 @@
+Integrating an external tool can be tricky!
+Luckily, the Tilt API allows us to quickly introspect what's happening behind the scenes.
+
+> 💡 Use `tilt describe kapp` to get a human readable version!
```
IMO this should go after the code snippet because before it, I don't know what you're talking about.

ellenkorbes commented on blog/_posts/2021-11-22-k8s-custom-deploy.md:
````diff
@@ -0,0 +1,123 @@
+Clearing temporary build folder: ./build/go-fn/
+Preparing: ./go-fn/ build/go-fn/function
+```
+In this case, our `apply_cmd` wrote back invalid YAML to `stdout` because we forgot to redirect the diagnostic logs to `stderr` - oops!
````
How would I fix this??

ellenkorbes commented:
>  blog/assets/images/k8s-custom-deploy/title.jpg

![image](https://user-images.githubusercontent.com/25550091/143046670-c30c71c3-38e5-40a3-8eec-80f6c45ce253.png)

nicks commented:
is this waiting on anything to go out?

milas commented:
> is this waiting on anything to go out?

No, this should have gone out last week...my fault, I made tweaks from L's feedback and never pushed.

Publishing momentarily!
